main/src/core/MonoDevelop.Core/MonoDevelop.Projects/WorkspaceObject.cs

```diff
@@ -98,15 +98,30 @@ protected void AssertMainThread ()
 		/// <typeparam name="T">Task return type</typeparam>
 		public Task<T> BindTask<T> (Func<CancellationToken, Task<T>> f)
 		{
-			var t = f (disposeCancellation.Token);
+			var taskCompletionSource = new TaskCompletionSource<bool> ();
 			lock (activeTasks) {
+				// Need to register a task before the Func is called otherwise the Dispose could be called whilst the
+				// Func's Task is being run and before the Func's task is registered.
 				if (disposeCancellation.IsCancellationRequested)
 					return Task.FromCanceled<T> (disposeCancellation.Token);
-				activeTasks.Add (t);
+				activeTasks.Add (taskCompletionSource.Task);
+			}
+			Task<T> t;
+			try {
+				t = f (disposeCancellation.Token);
+			} catch (Exception) {
+				// Do not prevent Project.OnDispose being called by leaving the task on the activeTasks list.
+				lock (activeTasks) {
+					taskCompletionSource.SetResult (true);
+					activeTasks.Remove (taskCompletionSource.Task);
+				}
+				throw;
```

Comment:
Okay, to clarify my poor review comment, instead of throwing here:
```suggestion
				return Task.FromException (e);
```